### PR TITLE
Support withGhEnv step

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Refactor `gh` in favour of `withGhEnv` that provides the installation and the environment to run the given closure.

## Why is it important?

Support for `closures`

### CODE REVIEW

File: vars/gh.groovy [hunk 1]
```diff
@@ -56,39 +52,23 @@ def call(Map args = [:]) {
...
-          } else {
-            if (v) {
-              flagsCommand += "--${k}='${normalise(v)}' "
-            }
+  withGhEnv(args) {
+    def flagsCommand = ''
+    if (flags) {
+      flags.each { k, v ->
+        log(level: 'DEBUG', text: "gh: k ${k} - v ${v}")
+        if (v instanceof java.util.ArrayList || v instanceof List) {
```
Comment (kuisathaverat) [inline vars/gh.groovy:60]: java.util.ArrayList implements List

File: vars/withGhEnv.groovy [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  return false
+}
+
+def downloadWithCurl(tarball, url) {
+  if(isInstalled(tool: 'curl', flag: '--version')) {
+    sh(label: 'download gsutil', script: "curl -sSLo ${tarball} --retry 3 --retry-delay 2 --max-time 10 ${url}")
+  } else {
+    log(level: 'WARN', text: 'withGhEnv: curl is not available. gh will not be installed then.')
+  }
+}
```
Comment (kuisathaverat) [inline vars/withGhEnv.groovy:91]: those functions are reusables as steps, WDYT?

**Review by kuisathaverat — APPROVED** [review]

File: vars/gh.groovy (hunk 1 above)
Comment (v1v) [reply vars/gh.groovy:60]: I don't recall now why I added it, but we could give a try, bear in mind, this change is about moving the installation to a different step. So I almost added nothing but refactoring.

File: vars/withGhEnv.groovy (hunk 2 above)
Comment (v1v) [reply vars/withGhEnv.groovy:91]: I like that, do you think we can refactor them in a follow up?

**Review by mdelapenya — APPROVED** [review]

File: vars/withGhEnv.groovy [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+@Field def ghLocation = ''
+
+def call(Map args = [:], Closure body) {
+  def credentialsId = args.get('credentialsId', '2a9602aa-ab9f-4e52-baf3-b71ca88469c7')
+  def version = args.get('version', '1.9.2')
+  def forceInstallation = args.get('forceInstallation', false)
+
+  if (ghLocation?.trim()) {
+    log(level: 'DEBUG', text: 'gh: get the ghLocation from cache.')
```
Comment (v1v) [inline vars/withGhEnv.groovy:36]:
```suggestion
    log(level: 'DEBUG', text: 'withGhEnv: get the ghLocation from cache.')
```

File: vars/withGhEnv.groovy [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  } else {
+    log(level: 'DEBUG', text: 'gh: set the ghLocation.')
```
Comment (v1v) [inline vars/withGhEnv.groovy:38]:
```suggestion
    log(level: 'DEBUG', text: 'withGhEnv: set the ghLocation.')
```

File: src/test/groovy/WithGhEnvStepTests.groovy [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+      ret = true
+    }
+    script.call() {
+      ret = true
+    }
+    printCallStack()
+    assertTrue(assertMethodCallContainsPattern('withEnv', 'PATH+GH'))
+    assertTrue(assertMethodCallContainsPattern('sh', 'wget -q -O'))
+    assertTrue(assertMethodCallContainsPattern('log', 'gh: get the ghLocation from cache.'))
+    assertTrue(assertMethodCallContainsPattern('log', 'gh: set the ghLocation.'))
```
Comment (v1v) [inline src/test/groovy/WithGhEnvStepTests.groovy:136]:
```suggestion
    assertTrue(assertMethodCallContainsPattern('log', 'withGhEnv: get the ghLocation from cache.'))
    assertTrue(assertMethodCallContainsPattern('log', 'withGhEnv: set the ghLocation.'))
```

**Pull request merged**